packages/medusa/src/repositories/product.ts

```diff
@@ -1,21 +1,408 @@
+        (queryBuilder, alias, topLevel) => {
+          if (topLevel === "categories") {
+            const joinScope =
+              where.categories as FindOptionsWhere<ProductCategory>
+
+            this._applyCategoriesQuery(queryBuilder, {
+              alias,
+              categoryAlias: "categories",
+              where: joinScope,
+              joinName: "leftJoinAndSelect",
+            })
+
+            return false
+          }
+
+          return true
+        },
+      ]
+    )
+  },
+
+  async findWithRelationsAndCount(
+    relations: string[] = [],
+    idsOrOptionsWithoutRelations: FindWithoutRelationsOptions = { where: {} }
+  ): Promise<[Product[], number]> {
+    const isOptionsArray = Array.isArray(idsOrOptionsWithoutRelations)
+    const originalWhere = isOptionsArray
+      ? undefined
+      : cloneDeep(idsOrOptionsWithoutRelations.where)
+    const originalOrder: any = isOptionsArray
+      ? undefined
+      : { ...idsOrOptionsWithoutRelations.order }
+    const clonedOptions = isOptionsArray
+      ? idsOrOptionsWithoutRelations
+      : cloneDeep(idsOrOptionsWithoutRelations)
+
+    let count: number
+    let entities: Product[]
+
+    if (isOptionsArray) {
+      entities = await this.find({
+        where: {
+          id: In(clonedOptions as unknown as string[]),
+        },
+        withDeleted: clonedOptions.withDeleted ?? false,
+      })
+      count = entities.length
+    } else {
+      const result = await this.queryProducts(clonedOptions, true)
+      entities = result[0]
+      count = result[1]
+    }
+    const entitiesIds = entities.map(({ id }) => id)
+
+    if (entitiesIds.length === 0) {
+      // no need to continue
+      return [[], count]
+    }
+
+    if (relations.length === 0) {
+      // Since we are finding by the ids that have been retrieved above and those ids are already
+      // applying skip/take. Remove those options to avoid getting no results
+      delete clonedOptions.skip
+      delete clonedOptions.take
+
+      const toReturn = await this.find({
+        ...(isOptionsArray
+          ? {}
+          : (clonedOptions as FindWithoutRelationsOptions)),
+        where: {
+          id: In(entitiesIds),
+          ...(isOptionsArray ? {} : clonedOptions.where),
+        },
+      })
+      return [toReturn, toReturn.length]
+    }
+
+    const groupedRelations = getGroupedRelations(relations)
+
+    const entitiesIdsWithRelations = await this.queryProductsWithIds(
+      entitiesIds,
+      groupedRelations,
+      idsOrOptionsWithoutRelations.withDeleted,
+      isOptionsArray
+        ? undefined
+        : objectToStringPath(idsOrOptionsWithoutRelations.select),
+      originalOrder,
+      originalWhere
+    )
+
+    const entitiesAndRelations = groupBy(entitiesIdsWithRelations, "id")
+    const entitiesToReturn = map(entitiesIds, (id) =>
+      merge({}, ...entitiesAndRelations[id])
+    )
+
+    return [entitiesToReturn, count]
+  },
+
+  async findWithRelations(
```

Comment:
thought(non-blocking): `findWithRelations` and `findWithRelationsAndCount` have very similar method bodies. Would it make sense to refactor some of the stuff to private helpers?